File: mobile/frontend/lib/data/repositories/emplacement_repository.dart
import 'package:frontend/data/models/emplacement.dart';
import 'base_repository.dart';

class EmplacementRepository extends BaseRepository<Emplacement> {
  EmplacementRepository() : super('emplacements');

  @override
  Emplacement fromMap(Map<String, dynamic> map) => Emplacement.fromMap(map);

  @override
  Map<String, dynamic> toMap(Emplacement entity) => entity.toMap();

  // ── Domain queries ────────────────────────────────────────

  /// Get all emplacements for a specific floor.
  Future<List<Emplacement>> getByFloor(int floor) {
    return query(where: 'floor = ?', whereArgs: [floor], orderBy: 'y ASC, x ASC');
  }

  /// Get a single emplacement by coordinates.
  Future<Emplacement?> getByCoords(int x, int y, int z, int floor) async {
    final results = await query(
      where: 'x = ? AND y = ? AND z = ? AND floor = ?',
      whereArgs: [x, y, z, floor],
      limit: 1,
    );
    return results.isEmpty ? null : results.first;
  }

  /// Get all storage slots (non-obstacle, non-road).
  Future<List<Emplacement>> getSlots({int? floor}) {
    if (floor != null) {
      return query(
        where: 'is_slot = 1 AND floor = ?',
        whereArgs: [floor],
      );
    }
    return query(where: 'is_slot = 1');
  }

  /// Get occupied slots.
  Future<List<Emplacement>> getOccupiedSlots({int? floor}) {
    if (floor != null) {
      return query(
        where: 'is_occupied = 1 AND floor = ?',
        whereArgs: [floor],
      );
    }
    return query(where: 'is_occupied = 1');
  }

  /// Get all emplacements containing a specific product.
  Future<List<Emplacement>> getByProduct(String productId) {
    return query(
      where: 'product_id = ? AND quantity > 0',
      whereArgs: [productId],
    );
  }

  /// Get available (empty) slots on a floor.
  Future<List<Emplacement>> getAvailableSlots(int floor) {
    return query(
      where: 'is_slot = 1 AND is_occupied = 0 AND floor = ?',
      whereArgs: [floor],
    );
  }

  /// Get all elevator cells.
  Future<List<Emplacement>> getElevators() {
    return query(where: 'is_elevator = 1');
  }

  /// Get expedition zones.
  Future<List<Emplacement>> getExpeditionZones() {
    return query(where: 'is_expedition = 1');
  }

  /// Get distinct floor numbers.
  Future<List<int>> getFloorNumbers() async {
    final db = await database;
    final results = await db.rawQuery(
      'SELECT DISTINCT floor FROM emplacements WHERE is_deleted = 0 ORDER BY floor ASC',
    );
    return results.map((r) => r['floor'] as int).toList();
  }

  /// Get grid dimensions for a floor.
  Future<Map<String, int>> getFloorDimensions(int floor) async {
    final db = await database;
    final result = await db.rawQuery(
      'SELECT MAX(x) as max_x, MAX(y) as max_y FROM emplacements WHERE floor = ? AND is_deleted = 0',
      [floor],
    );
    if (result.isEmpty) return {'width': 0, 'height': 0};
    return {
      'width': ((result.first['max_x'] as int?) ?? 0) + 1,
      'height': ((result.first['max_y'] as int?) ?? 0) + 1,
    };
  }

  /// Update stock at a specific emplacement.
  Future<void> updateStock(String emplacementId, String? productId, int quantity) {
    return update(emplacementId, {
      'product_id': productId,
      'quantity': quantity,
      'is_occupied': (productId != null && quantity > 0) ? 1 : 0,
    });
  }
}

File: mobile/backend/app/repositories/emplacement_repository.py
"""
Emplacement repository for Firestore operations.
"""

from typing import List, Optional, Dict, Any

from app.repositories.base_repository import BaseRepository


class EmplacementRepository(BaseRepository):
    """Repository for Emplacement location documents."""

    def __init__(self):
        super().__init__("emplacements")

    async def get_by_coordinates(
        self, x: int, y: int, z: int = 0, floor: int = 0
    ) -> Optional[Dict[str, Any]]:
        """Find a location by exact coordinates."""
        results = await self.query(filters=[
            ("x", "==", x),
            ("y", "==", y),
            ("z", "==", z),
            ("floor", "==", floor),
        ])
        return results[0] if results else None

    async def get_slots_by_floor(self, floor: int) -> List[Dict[str, Any]]:
        """Get all storage slots on a specific floor."""
        return await self.query(filters=[
            ("floor", "==", floor),
            ("is_slot", "==", True),
        ])

    async def get_available_slots(self, floor: Optional[int] = None) -> List[Dict[str, Any]]:
        """Get available (unoccupied) storage slots."""
        filters = [("is_slot", "==", True), ("is_occupied", "==", False)]
        if floor is not None:
            filters.append(("floor", "==", floor))
        return await self.query(filters=filters)

    async def get_occupied_slots(self, floor: Optional[int] = None) -> List[Dict[str, Any]]:
        """Get occupied storage slots."""
        filters = [("is_slot", "==", True), ("is_occupied", "==", True)]
        if floor is not None:
            filters.append(("floor", "==", floor))
        return await self.query(filters=filters)

    async def get_expedition_zones(self) -> List[Dict[str, Any]]:
        """Get all expedition zones."""
        return await self.query(filters=[("is_expedition", "==", True)])

    async def get_product_locations(self, product_id: str) -> List[Dict[str, Any]]:
        """Get all locations storing a specific product."""
        return await self.query(filters=[("product_id", "==", product_id)])

    async def get_filtered(
        self,
        floor: Optional[int] = None,
        is_slot: Optional[bool] = None,
        is_occupied: Optional[bool] = None,
    ) -> List[Dict[str, Any]]:
        """Get locations with optional filters."""
        filters = []
        if floor is not None:
            filters.append(("floor", "==", floor))
        if is_slot is not None:
            filters.append(("is_slot", "==", is_slot))
        if is_occupied is not None:
            filters.append(("is_occupied", "==", is_occupied))
        return await self.query(filters=filters if filters else None)
